src/ControlSystem/Averager.hpp

```diff
@@ -0,0 +1,77 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <array>
+#include <cstddef>
+#include <deque>
+
+#include "DataStructures/DataVector.hpp"
+
+/// \ingroup ControlSystemGroup
+/// A weighted exponential averager of Q and its derivatives
+/// implementing Appendix A in Hemberger et al, https://arxiv.org/abs/1211.6079
+/// The averager is designed to support an arbitrary DerivOrder, however the
+/// private get_derivs method currently only supports DerivOrder = 2.
+/// If an additional DerivOrder is needed, finite differencing needs to be
+/// implemented to specifically handle that order (as it seems like overkill to
+/// generalize the differencing stencil at this time).
+template <size_t DerivOrder>
+class Averager {
+ public:
+  /// `avg_timescale_frac` determines the exponential averaging timescale
+  /// through \f$\tau_\mathrm{avg} = \f$`avg_timescale_frac`\f$\times \tau\f$,
+  /// where \f$\tau\f$ is the damping time.
+  /// `avg_timescale_frac` must be positive.
+  /// `average_0th_deriv_of_q` determines whether the call operator returns an
+  /// averaged or unaveraged quantity for the \f$Q\f$ function values.
+  /// `true` returns an averaged \f$Q\f$. `false` returns the raw \f$Q\f$.
+  /// The derivatives are always averaged, so `average_0th_deriv_of_q` affects
+  /// only the function value piece of the returned array.
+  Averager(double avg_timescale_frac, bool average_0th_deriv_of_q) noexcept;
+
+  /// Returns the averaged/unaveraged (`average_0th_deriv_of_q`=`true`/`false`)
+  /// values of \f$Q\f$ and averaged derivatives of \f$Q\f$ up to
+  /// the `DerivOrder`'th derivative, at \f$t=\f$`time`.
+  std::array<DataVector, DerivOrder + 1> operator()(double time) const noexcept;
+  /// A function that allows for resetting the averager.
+  void clear() noexcept;
+  /// Checks whether the averager can provide averaged data at \f$t=\f$`time`.
+  bool have_valid_data(double time) const noexcept {
```

Comment:
I have attempted to clarify the purpose of this function in the new commits. Essentially, it is just checking that there is sufficient data to provide numerical derivatives for the specified DerivOrder (and would still be necessary even if we did not average the derivatives). I think the fact that the 'Averager' does more than just average is probably confusing -- especially since its main purpose is to provide derivatives and the averaging part is just used to provide better approximations of the derivatives...